staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go

```diff
@@ -235,39 +212,29 @@ func (r *crdHandler) removeDeadStorage() {
 			}
 		}
 		if !found {
-			delete(storageMap, uid)
+			delete(r.customStorage, uid)
 		}
 	}
-
-	r.customStorageLock.Lock()
-	defer r.customStorageLock.Unlock()
-
-	r.customStorage.Store(storageMap)
 }
 
 // GetCustomResourceListerCollectionDeleter returns the ListerCollectionDeleter for
 // the given uid, or nil if one does not exist.
-func (r *crdHandler) GetCustomResourceListerCollectionDeleter(uid types.UID) finalizer.ListerCollectionDeleter {
-	storageMap := r.customStorage.Load().(crdStorageMap)
-	ret, ok := storageMap[uid]
-	if !ok {
-		return nil
-	}
-	return ret.storage
+func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions.CustomResourceDefinition) finalizer.ListerCollectionDeleter {
+	info := r.getServingInfoFor(crd)
+	return info.storage
 }
 
 func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefinition) *crdInfo {
-	storageMap := r.customStorage.Load().(crdStorageMap)
-	ret, ok := storageMap[crd.UID]
+	r.customStorageLock.RLock()
```

Comment:
@liggitt specifically requested an atomic.Value is several spots with the claim that it's read performance was better than grabbing a non-contentious read lock.  I don't know one way or the other, but it was an intentional choice.